design/feature-flags.md

```diff
@@ -0,0 +1,98 @@
+# "Future" Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem, of course, is that
+breaking changes are only allowed in major versions and those are rare.
```

Comment:
Devil's advocate here. What is wrong with following SemVer and why must major versions be rare for this young repository?

I think most of the issues come from affecting existing cloud resources which may result from Amazon CDK breaking its API. It might be helpful to think about how we can reduce the impact on existing cloud resources when we inevitably introduce new major versions. Here are a few off the top of my mind:

- Migration guides
- Scripts to compare CloudFormation template diffs across versions to determine what upgrades are safe
- Make it very visible if someone wants to deploy a new CloudFormation template right after a major upgrade

Reply:
Major version releases are expensive and hard and require quite a lot of effort both for us and for users. We generally rather make the effort of not breaking users as much as possible and reduce the amount of major version chrun (see k8s), but figuing out a decent way to still move forward when needed. This pattern is a way to allow both for agility (albeit at some expense) and stability to users.